### PR TITLE
[AUTO] Update Auto-device plugin selection logic

### PR DESCRIPTION
**Pull request by coneypo**
### Details:
 - *Auto-device plugin selects device with the priority of dGPU, VPUX, iGPU, MYRIAD and CPU*
 - *Based on [#6113 ](https://github.com/openvinotoolkit/openvino/pull/6113), add tests into myraidFuncTests with Auto-devie plugin*

### Tickets:
 - *58005*

### CODE REVIEW

**Comment by coneypo** [comment]
Will add tests for myriad with auto-device plugin once https://github.com/openvinotoolkit/openvino/pull/6113 ready;

**Comment by myshevts** [comment]
as a draft (e.g. for the validation purposes) this is fine, but in the actual product the MYX should have very low priority.
unless it is the only device in the system (RaspPi + MYX stick, as we have disucssed with @wangleis )

File: inference-engine/src/auto_plugin/auto_plugin.cpp [hunk 1]
```diff
@@ -233,6 +233,7 @@ DeviceName AutoInferencePlugin::SelectDevice(const std::vector<DeviceName>& meta
 
     std::vector<DeviceName> CPU;
     std::vector<DeviceName> GPU;
+    std::vector<DeviceName> VPU;
```
Comment (mashoujiang) [inline inference-engine/src/auto_plugin/auto_plugin.cpp:236]: better rename to "MYRIAD", also for the PR title name.

**Review by mashoujiang — COMMENTED** [review]
Need another commit to add tests for myraid

File: inference-engine/src/auto_plugin/auto_plugin.cpp (hunk 1 above)
Comment (coneypo) [reply inference-engine/src/auto_plugin/auto_plugin.cpp]: VPUX > dGPU > iGPU > CPU > MYRIAD now

**Comment by coneypo** [comment]
> as a draft (e.g. for the validation purposes) this is fine, but in the actual product the MYX should have very low priority.
> unless it is the only device in the system (RaspPi + MYX stick, as we have disucssed with @wangleis )

Update the selection device logic to **VPUX > dGPU > iGPU > CPU > MYRIAD** now, will try VPUX once our kmb HW ready;

File: inference-engine/src/auto_plugin/auto_plugin.cpp [hunk 2]
```diff
@@ -274,37 +274,82 @@ DeviceName AutoInferencePlugin::SelectDevice(const std::vector<DeviceName>& meta
...
+            auto supportNetwork = std::find(capability.begin(), capability.end(), networkPrecision);
+            if (supportNetwork != capability.end()) {
+                return item;
+            }
+        }
+    } else if (!MYRIAD.empty()) {
+        for (auto&& item : MYRIAD) {
+            std::vector<std::string> capability = GetCore()->GetMetric(item, METRIC_KEY(OPTIMIZATION_CAPABILITIES));
+            // If this device supports "FP16" in "OPTIMIZATION_CAPABILITIES", supports FP32 at the same time (Currently for MYRIAD only)
+            auto supportFP16 = std::find(capability.begin(), capability.end(), "FP16");
```
Comment (coneypo) [inline inference-engine/src/auto_plugin/auto_plugin.cpp:339]: For Myriad device , `OPTIMIZATION_CAPABILITIES: FP16` only, but we can load FP32 network to it;

**Comment by coneypo** [comment]
@ilya-lavrenov  @myshevts @wangleis @mashoujiang CI is green now, please review this PR to support Myriad in auto-device plugin;

File: inference-engine/src/auto_plugin/auto_plugin.cpp [hunk 3]
```diff
@@ -274,34 +274,78 @@ DeviceName AutoInferencePlugin::SelectDevice(const std::vector<DeviceName>& meta
...
+    std::vector<DeviceName> VPUX;
 
     for (auto& item : metaDevices) {
         if (item.find("CPU") == 0) {
             CPU.push_back(item);
             continue;
         }
-        if (item.find("GPU") == 0) {
-            GPU.push_back(item);
+        if (item.find("GPU.1") == 0) {
```
Comment (mashoujiang) [inline inference-engine/src/auto_plugin/auto_plugin.cpp:287]: what about GPU.2?

File: inference-engine/src/auto_plugin/auto_plugin.cpp [hunk 4]
```diff
@@ -274,34 +274,78 @@ DeviceName AutoInferencePlugin::SelectDevice(const std::vector<DeviceName>& meta
...
+            // dGPU found
+            dGPU.push_back(item);
+            continue;
+        } else if (item.find("GPU.0") == 0 || item.find("GPU") == 0) {
+            // iGPU found
```
Comment (mashoujiang) [inline inference-engine/src/auto_plugin/auto_plugin.cpp:292]: this GPU.0 or GPU also may be dGPU

File: inference-engine/src/auto_plugin/auto_plugin.cpp [hunk 5]
```diff
@@ -274,34 +274,78 @@ DeviceName AutoInferencePlugin::SelectDevice(const std::vector<DeviceName>& meta
...
+                return item;
+            }
+        }
+    } else if (!MYRIAD.empty()) {
+        for (auto&& item : MYRIAD) {
+            std::vector<std::string> capability = GetCore()->GetMetric(item, METRIC_KEY(OPTIMIZATION_CAPABILITIES));
+            // If this device supports "FP16" in "OPTIMIZATION_CAPABILITIES", supports FP32 at the same time (Currently for MYRIAD only)
+            auto supportFP16 = std::find(capability.begin(), capability.end(), "FP16");
+            if (supportFP16 != capability.end()) {
+                capability.insert(capability.begin(), "FP32");
```
Comment (mashoujiang) [inline inference-engine/src/auto_plugin/auto_plugin.cpp:341]: what about directly check supportNetwork == FP32 if no capability support.
It is not good way to insert element to a vector, and why insert to begin()? why not directly push_back()?

**Review by mashoujiang — COMMENTED** [review]
the logic is not right now, if only dGPU is available, the device will be "GPU" or "GPU.0".

File: inference-engine/src/auto_plugin/auto_plugin.cpp (hunk 4 above)
Comment (coneypo) [reply inference-engine/src/auto_plugin/auto_plugin.cpp]: Will check device full name to update GPU device list: `Intel(R) Iris(R) Xe Graphics (iGPU)` and `Intel(R) Iris(R) Xe MAX Graphics (dGPU)` for example;

**Comment by coneypo** [comment]
If network is FP32 but there is no device support FP32, will loop the device list and offload FP32 network to device support FP16.

**Review by mashoujiang — APPROVED** [review]
LGTM, it is aligned with jira ticket.

**Comment by coneypo** [comment]
@myshevts Hi, Maxim, please help review this PR for updating selecting logic in auto-device plugin, with myriad and vpux added, thanks:)

**Review by myshevts — APPROVED** [review]

**Pull request merged**